{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "cwd = os.getcwd()\n",
    "pygad_loc = '/grid/home/nbourgeois/codonOpt'\n",
    "os.chdir(pygad_loc)\n",
    "from general_functions import *\n",
    "from metrics import *\n",
    "os.chdir(cwd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ga_input = '/grid/home/nbourgeois/data/codon_jason/final_seq.fa'\n",
    "output = '' #output filename\n",
    "tissue = 'Liver'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_codon_weights(loc):\n",
    "    # read table\n",
    "    try:\n",
    "        index_loc = os.path.join(pygad_loc,loc)\n",
    "        codon_table_raw = pd.read_csv(index_loc,sep='\\t')\n",
    "\n",
    "    except FileNotFoundError:\n",
    "        print('Invalid Tissue')\n",
    "        return()    \n",
    "\n",
    "    codon_table = codon_table_raw.set_index('#Codon')\n",
    "    maxes = codon_table_raw.groupby(by=['AA'])['Frequency'].max()\n",
    "    codon_table['max_col'] = codon_table.apply(lambda x: maxes[x['AA']],axis=1)\n",
    "    codon_table['weight'] = codon_table['Frequency'] / codon_table['max_col']\n",
    "    weight_dict = codon_table['weight'].to_dict()\n",
    "    \n",
    "    return(weight_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "(fastaIDs, fastaSeqs) = readFasta(ga_input)\n",
    "names = ['_'.join(anid.split(':')[1].split('_')[0:2]) for anid in fastaIDs]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CAI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "orig_weights = get_codon_weights('references/CoCoPUTs_codon_usage/codon_usage/Liver.codon.txt')\n",
    "# bai_weight_dict = get_bicodon_weights(tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_weights = get_codon_weights('tissue_cocoputs/data/Test_liver.codon.txt')\n",
    "hi_weights = get_codon_weights('tissue_cocoputs/data/Test_liver_high.codon.txt')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "orig_cais = []\n",
    "new_cais = []\n",
    "hi_cais = []\n",
    "diffs = []\n",
    "hidiffs = []\n",
    "for fid,aa_seq in zip(fastaIDs, fastaSeqs):\n",
    "    orig_cai = get_cai(aa_seq, orig_weights) *100\n",
    "    orig_cais.append(orig_cai)\n",
    "    new_cai = get_cai(aa_seq, new_weights)*100\n",
    "    new_cais.append(new_cai)\n",
    "    hi_cai = get_cai(aa_seq, hi_weights)*100\n",
    "    hi_cais.append(hi_cai)\n",
    "    diffs.append(new_cai - orig_cai)\n",
    "    hidiffs.append(hi_cai - orig_cai)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Orig</th>\n",
       "      <th>New</th>\n",
       "      <th>Diff</th>\n",
       "      <th>High</th>\n",
       "      <th>Hi-Diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cai_0</td>\n",
       "      <td>83.963608</td>\n",
       "      <td>83.890579</td>\n",
       "      <td>-0.073029</td>\n",
       "      <td>80.210590</td>\n",
       "      <td>-3.753018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cai_1</td>\n",
       "      <td>88.048398</td>\n",
       "      <td>88.262931</td>\n",
       "      <td>0.214533</td>\n",
       "      <td>85.284358</td>\n",
       "      <td>-2.764040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cai_2</td>\n",
       "      <td>87.765231</td>\n",
       "      <td>87.993236</td>\n",
       "      <td>0.228005</td>\n",
       "      <td>84.871991</td>\n",
       "      <td>-2.893240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cpg_0</td>\n",
       "      <td>66.546848</td>\n",
       "      <td>67.038484</td>\n",
       "      <td>0.491636</td>\n",
       "      <td>60.166190</td>\n",
       "      <td>-6.380658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cpg_1</td>\n",
       "      <td>65.014299</td>\n",
       "      <td>65.551462</td>\n",
       "      <td>0.537163</td>\n",
       "      <td>58.606760</td>\n",
       "      <td>-6.407539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cpg_2</td>\n",
       "      <td>68.012043</td>\n",
       "      <td>68.466356</td>\n",
       "      <td>0.454314</td>\n",
       "      <td>62.004401</td>\n",
       "      <td>-6.007642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>bai_0</td>\n",
       "      <td>87.985176</td>\n",
       "      <td>88.135984</td>\n",
       "      <td>0.150808</td>\n",
       "      <td>84.877158</td>\n",
       "      <td>-3.108018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>bai_1</td>\n",
       "      <td>88.680959</td>\n",
       "      <td>88.947771</td>\n",
       "      <td>0.266812</td>\n",
       "      <td>85.516224</td>\n",
       "      <td>-3.164735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>bai_2</td>\n",
       "      <td>87.944822</td>\n",
       "      <td>88.121900</td>\n",
       "      <td>0.177078</td>\n",
       "      <td>84.480205</td>\n",
       "      <td>-3.464617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cpg-n_0</td>\n",
       "      <td>59.220453</td>\n",
       "      <td>57.606388</td>\n",
       "      <td>-1.614065</td>\n",
       "      <td>53.373740</td>\n",
       "      <td>-5.846713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cpg-n_1</td>\n",
       "      <td>57.894029</td>\n",
       "      <td>56.197872</td>\n",
       "      <td>-1.696157</td>\n",
       "      <td>51.801956</td>\n",
       "      <td>-6.092073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>cpg-n_2</td>\n",
       "      <td>59.553181</td>\n",
       "      <td>57.896875</td>\n",
       "      <td>-1.656307</td>\n",
       "      <td>53.614785</td>\n",
       "      <td>-5.938396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>bai-n_0</td>\n",
       "      <td>46.951676</td>\n",
       "      <td>45.950458</td>\n",
       "      <td>-1.001218</td>\n",
       "      <td>40.940912</td>\n",
       "      <td>-6.010763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>bai-n_1</td>\n",
       "      <td>46.947887</td>\n",
       "      <td>45.973036</td>\n",
       "      <td>-0.974851</td>\n",
       "      <td>40.884932</td>\n",
       "      <td>-6.062955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>bai-n_2</td>\n",
       "      <td>47.175597</td>\n",
       "      <td>46.246843</td>\n",
       "      <td>-0.928754</td>\n",
       "      <td>41.046878</td>\n",
       "      <td>-6.128718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>cai-n_0</td>\n",
       "      <td>50.608221</td>\n",
       "      <td>50.209123</td>\n",
       "      <td>-0.399098</td>\n",
       "      <td>44.702472</td>\n",
       "      <td>-5.905750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>cai-n_1</td>\n",
       "      <td>56.511879</td>\n",
       "      <td>56.310627</td>\n",
       "      <td>-0.201252</td>\n",
       "      <td>50.535220</td>\n",
       "      <td>-5.976658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>cai-n_2</td>\n",
       "      <td>50.563666</td>\n",
       "      <td>49.854595</td>\n",
       "      <td>-0.709071</td>\n",
       "      <td>44.352790</td>\n",
       "      <td>-6.210877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>RandomSeq_0</td>\n",
       "      <td>64.642662</td>\n",
       "      <td>64.744531</td>\n",
       "      <td>0.101869</td>\n",
       "      <td>58.764135</td>\n",
       "      <td>-5.878527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RandomSeq_1</td>\n",
       "      <td>60.473684</td>\n",
       "      <td>60.535776</td>\n",
       "      <td>0.062092</td>\n",
       "      <td>54.637991</td>\n",
       "      <td>-5.835694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>RandomSeq_2</td>\n",
       "      <td>62.808877</td>\n",
       "      <td>62.476142</td>\n",
       "      <td>-0.332735</td>\n",
       "      <td>57.060888</td>\n",
       "      <td>-5.747989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>PoorSplicing_0</td>\n",
       "      <td>62.178381</td>\n",
       "      <td>61.756598</td>\n",
       "      <td>-0.421783</td>\n",
       "      <td>56.011551</td>\n",
       "      <td>-6.166829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PoorSplicing_1</td>\n",
       "      <td>63.368128</td>\n",
       "      <td>63.522791</td>\n",
       "      <td>0.154663</td>\n",
       "      <td>57.809696</td>\n",
       "      <td>-5.558432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PoorSplicing_2</td>\n",
       "      <td>64.085568</td>\n",
       "      <td>64.192931</td>\n",
       "      <td>0.107364</td>\n",
       "      <td>58.911411</td>\n",
       "      <td>-5.174157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PoorSplicing_3</td>\n",
       "      <td>63.063004</td>\n",
       "      <td>62.992158</td>\n",
       "      <td>-0.070845</td>\n",
       "      <td>57.415554</td>\n",
       "      <td>-5.647449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PoorSplicing_4</td>\n",
       "      <td>61.777497</td>\n",
       "      <td>61.722880</td>\n",
       "      <td>-0.054618</td>\n",
       "      <td>55.755714</td>\n",
       "      <td>-6.021784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PoorSplicing_5</td>\n",
       "      <td>59.278227</td>\n",
       "      <td>59.086308</td>\n",
       "      <td>-0.191918</td>\n",
       "      <td>53.480994</td>\n",
       "      <td>-5.797233</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Name       Orig        New      Diff       High   Hi-Diff\n",
       "0            cai_0  83.963608  83.890579 -0.073029  80.210590 -3.753018\n",
       "1            cai_1  88.048398  88.262931  0.214533  85.284358 -2.764040\n",
       "2            cai_2  87.765231  87.993236  0.228005  84.871991 -2.893240\n",
       "3            cpg_0  66.546848  67.038484  0.491636  60.166190 -6.380658\n",
       "4            cpg_1  65.014299  65.551462  0.537163  58.606760 -6.407539\n",
       "5            cpg_2  68.012043  68.466356  0.454314  62.004401 -6.007642\n",
       "6            bai_0  87.985176  88.135984  0.150808  84.877158 -3.108018\n",
       "7            bai_1  88.680959  88.947771  0.266812  85.516224 -3.164735\n",
       "8            bai_2  87.944822  88.121900  0.177078  84.480205 -3.464617\n",
       "9          cpg-n_0  59.220453  57.606388 -1.614065  53.373740 -5.846713\n",
       "10         cpg-n_1  57.894029  56.197872 -1.696157  51.801956 -6.092073\n",
       "11         cpg-n_2  59.553181  57.896875 -1.656307  53.614785 -5.938396\n",
       "12         bai-n_0  46.951676  45.950458 -1.001218  40.940912 -6.010763\n",
       "13         bai-n_1  46.947887  45.973036 -0.974851  40.884932 -6.062955\n",
       "14         bai-n_2  47.175597  46.246843 -0.928754  41.046878 -6.128718\n",
       "15         cai-n_0  50.608221  50.209123 -0.399098  44.702472 -5.905750\n",
       "16         cai-n_1  56.511879  56.310627 -0.201252  50.535220 -5.976658\n",
       "17         cai-n_2  50.563666  49.854595 -0.709071  44.352790 -6.210877\n",
       "18     RandomSeq_0  64.642662  64.744531  0.101869  58.764135 -5.878527\n",
       "19     RandomSeq_1  60.473684  60.535776  0.062092  54.637991 -5.835694\n",
       "20     RandomSeq_2  62.808877  62.476142 -0.332735  57.060888 -5.747989\n",
       "21  PoorSplicing_0  62.178381  61.756598 -0.421783  56.011551 -6.166829\n",
       "22  PoorSplicing_1  63.368128  63.522791  0.154663  57.809696 -5.558432\n",
       "23  PoorSplicing_2  64.085568  64.192931  0.107364  58.911411 -5.174157\n",
       "24  PoorSplicing_3  63.063004  62.992158 -0.070845  57.415554 -5.647449\n",
       "25  PoorSplicing_4  61.777497  61.722880 -0.054618  55.755714 -6.021784\n",
       "26  PoorSplicing_5  59.278227  59.086308 -0.191918  53.480994 -5.797233"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(data = {'Name':names,'Orig':orig_cais,'New':new_cais,'Diff':diffs,'High':hi_cais,'Hi-Diff':hidiffs})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3441, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_131766/2365681367.py\", line 2, in <module>\n",
      "    cai = get_cai(aa_seq, cai_weight_dict)\n",
      "NameError: name 'cai_weight_dict' is not defined\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'NameError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3441, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_131766/2365681367.py\", line 2, in <module>\n",
      "    cai = get_cai(aa_seq, cai_weight_dict)\n",
      "NameError: name 'cai_weight_dict' is not defined\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'NameError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3361, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3458, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1125, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3441, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_131766/2365681367.py\", line 2, in <module>\n",
      "    cai = get_cai(aa_seq, cai_weight_dict)\n",
      "NameError: name 'cai_weight_dict' is not defined\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'NameError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3361, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3458, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1125, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2944, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3170, in run_cell_async\n",
      "    interactivity=interactivity, compiler=compiler, result=result)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3380, in run_ast_nodes\n",
      "    self.showtraceback()\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1143, in structured_traceback\n",
      "    chained_exceptions_tb_offset)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/data/software/anaconda3/envs/bfx-gpu/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for fid,aa_seq in zip(fastaIDs, fastaSeqs):\n",
    "    cai = get_cai(aa_seq, cai_weight_dict)\n",
    "\n",
    "    print(\"{}:\\tCAI is {}\".format(fid,cai))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BAI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#bai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR! Session/line number was not unique in database. History logging moved to new session 2789\n"
     ]
    },
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (1939985177.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"/tmp/ipykernel_131766/1939985177.py\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    for aa_seq,id in fastaIDs fastaSeqs:\u001b[0m\n\u001b[0m                                      ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "for aa_seq,id in fastaIDs fastaSeqs:\n",
    "    bai = get_bai(aa_seq, bai_weight_dict)\n",
    "    print(\"{}:\\tBAI is {}\".format(id,bai))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cpg\n",
    "for aa_seq,id in fastaIDs fastaSeqs:\n",
    "    bai = get_cpg(aa_seq, bai_weight_dict)\n",
    "    print(\"{}:\\tcpg is {}\".format(id,bai))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "global codon_to_int\n",
    "\n",
    "codon_to_int, gene_space = init_parameters_fun(aa_seq)\n",
    "    \n",
    "gene_space_int = [[codon_to_int[x] for x in y] for y in gene_space]"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "fbeaddbcfa20e556290e66709676010ba3fa91a169757ce7752c597d396ac62c"
  },
  "kernelspec": {
   "display_name": "Python 3 (GPU)",
   "language": "python",
   "name": "bfx-gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
